### DISCUSSION TITLE
Future plans for discord.py

### DISCUSSION THREAD

**Discussion by hkennyv** [Cog-Creators/Red-DiscordBot#5232] [Q&A] [answered]
Hi there,

Thank you to all the maintainers for the hard work. I just wanted to bring it to attention that it looks like development on [discord.py](https://github.com/Rapptz/discord.py) has halted (for some very good reasons outlined [here](https://gist.github.com/Rapptz/4a2f62751b9600a31a0d3c78100287f1)). As I understand it, that is the core API library this project uses.

I didn't see a discussion on it yet, so I apologize if this issue is misplaced. As a user, I would love to know what that means for this project and how I could possibly help in any way.

**Comment by Twentysix26** [comment]
Hi,
it is way too early to have an official stance on the matter :)
Consider following our announcement channel in the [discord server](https://discord.gg/red). If anything major needs to be announced it will be done there.

**Comment by Tobotimus** [comment]
Hi guys, it's been a while, but I thought I'd give my 2 cents here.

I'm assuming the discussion on what Red does going forward will centre around whether to fork discord.py, or whether to go with a different fork. Here's my opinion on that.

The idea of forking discord.py came up a number of times whilst I was maintaining this project, and I was always staunchly against it. I assume it also came up after I became inactive. From my recollection, the main reasons for people wanting to fork was not necessarily due to gripes with how the library was written or maintained, but just to add ways for us to hook into internal discord.py functionality. I always viewed this as a dangerously burdensome task with not a whole lot of payoff - discord.py is a large, well-written library, and it was clear that Danny had a vision on how to maintain it. It would have been a large responsibility to fork discord.py and add our own things to it, even if a lot of that work would have been the inevitable resolving of merge conflicts when trying to bring forth changes from upstream. Now imagine how much more responsibility it would be if there was no upstream, and the entire direction of the library was up to us.

I was the first to fork the `discord.ext.commands` package (in a way, effectively "shadowing" it as `redbot.core.commands`), because we needed a lot more control over command permissions for the `permissions` cog to work effectively. The `discord.ext.commands` package was the bot framework, whilst the core `discord` package was the wrapper for the Discord API, so as a bot framework ourselves, I think it made a lot of sense for us to have forked `discord.ext.commands` and take full control of it. 

At this point, the consequences of forking discord.py haven't really changed - however, the consequences of *not* forking discord.py *have* changed. My advice is essentially this: before deciding to fork, give it a bit of time to see what happens. I understand that there are some deadlines from Discord on when some changes need to be made (new privileges/verification/slash commands), so you'll need to have the timeline in mind - set yourselves an approximate deadline for yourselves to decide on whether to fork discord.py, or go with someone else's fork. Who knows, perhaps someone Danny trusts will change their mind and become the new maintainer of discord.py. If I were still maintaining Red, I wouldn't want to rush into taking on discord.py as Cog-Creator's responsibility, even if we agreed to only maintain the library for Red's purposes, and not for the wider community.

Best of luck with everything guys ❤

**Comment by lveneris** [comment]
With regards to forking discord.py, I believe it could present some good opportunities. Remember that d.py was maintained by one person - we'd have good community support to make this a team effort, and many hands make light work of heavy duties.

A good approach would be to factor out a lot of Red's improvements to `discord.ext.commands` out as a framework, similar to [the Sapphire project](https://github.com/sapphiredev/framework)'s relationship to `discord.js`. From there, whether the Red community entirely forks d.py is up to the masses, but at the very least it'd be wise to create a better official framework for wider use (and our use, dogfooding is always good) and get some higher ups in a position where input is valued in design decisions for emerging forks.

That's my two cents as someone that started with d.py and Red that moved over to discord.js to help them with Sapphire, became an external contributor, and stuck around for long enough to watch it become the official discord.js framework and many staff become collaborators of discord.js itself.

I believe if the community wanted to, it could pull it off.

**Comment by tigattack** [comment]
I understand this is an open source project backed by many individual contributors, and I _hugely_ appreciate the work of every contributor. This project has allowed me to run a number of personal and community bots which has made my life, and those of community members, much easier in many cases.

_However_... I think more discussion around this, and eventually a plan for how to progress forwards, is crucial.

The latest release of discord.py (1.7.3, also used by Red) talks to Discord API v7 (seen [here](https://github.com/Rapptz/discord.py/blob/v1.7.3/discord/http.py#L54)), which will be **decomissioned** on 30/04/2022.  
See here: https://github.com/discord/discord-api-docs/discussions/4510

From this date, all Red-based bots **will not function**.  
This is huge, and will likely be a surprise to many since there hasn't been much discussion around this.

I'm not requesting that the Red contributors fix this, but I do believe that a decision should be made as to how to progress, considering the incoming ramifications of inaction.  
Even if the decision is to just not do anything, I think this should be clear and should be communicated to users and community members.

**Reply by Jackenmen** [reply]
There are active plans to solve the problem due to the departure of discord.py. We just haven't taken an official stance on it yet. When we do, we'll announce it.

**Reply by Twentysix26** [reply]
Being worried in light of those news is understandable but if the plan was "do nothing" we'd already have stated so.
We have plans on how to deal with upcoming events and as a end user you shouldn't concern yourself with any of this, we have bots that need to live on too so it's in our best interest that things will still work :-)
If you develop cogs, you can expect them to still work after making some adjustments. For that, we're asking you to stay tuned, the path forward will be announced soon ™️

**Reply by tigattack** [reply]
This is very reassuring to hear. Thank you both!

**Comment by Kowlin** [comment, marked as answer]
*Cross posting from <https://discord.com/channels/133049272517001216/133080797660971008/949834797801422858>*

---

**Regarding the future of Red**

You may have seen Danny's recent announcement regarding his return to developing discord.py (https://gist.github.com/Rapptz/c4324f17a80c94776832430007ad40e6). We are pleased to announce that Red 3.5 will be continuing to use discord.py, and will have support for all of your favorite new discord features included on release.

**Notices for Third Party Developers**

A development version of Red 3.5 with discord.py 2.0 is available in the `V3/dpy2_unstable` branch. You can install this version by replacing the normal pip install command with
`python -m pip install -U --force-reinstall git+https://github.com/Cog-Creators/Red-DiscordBot@V3/dpy2_unstable#egg=Red-DiscordBot`
If you want to follow development for d.py 2 support or if you find any bugs with this branch, refer to PR #5600.

As 3.5 is a minor version bump, it is going to come with many breaking changes for third party cogs. Red has not finalized its changes for 3.5, however the discord.py 2.0 migration guide can be found at https://discordpy.readthedocs.io/en/master/migrating.html.

If you are for whatever reason unable to update your cogs for 3.5 before April 30th, but still want your cogs to be usable after that point, we are offering migration assistance at request. Please contact an org member for more information.

**In other news**

Red has hit 3000 stars on GitHub! We would like to thank all of you for your continued support.

**Reply by Kowlin** [reply]
CC: @tigattack, I'm sure you'll be interested in this.
